libcxx/include/streambuf

```diff
@@ -233,6 +234,9 @@ protected:
   inline _LIBCPP_HIDE_FROM_ABI_AFTER_V1 void gbump(int __n) { __ninp_ += __n; }
 
   inline _LIBCPP_HIDE_FROM_ABI_AFTER_V1 void setg(char_type* __gbeg, char_type* __gnext, char_type* __gend) {
+    _LIBCPP_ASSERT_VALID_INPUT_RANGE(__gbeg <= __gnext, "[gbeg, gnext) must be a valid range");
+    _LIBCPP_ASSERT_VALID_INPUT_RANGE(__gbeg <= __gend, "[gbeg, gend) must be a valid range");
+    _LIBCPP_ASSERT_VALID_INPUT_RANGE(__gnext <= __gend, "[gnext, gend) must be a valid range");
```

Comment:
Not related to this patch, but is this the best way to test this?
If it's not a valid range we trigger UB. Would it be better to test the value cased to an `uintptr_t`?

Reply:
I've thought about adding something similar. If we wanted to we could check with `__asan_region_is_poisoned` that inputs to algorithms (or here) are in fact valid memory. That would probably reduce the amount of not-actually-valid uses of algorithms in the wild quite a bit. e.g. with the `mismatch` vectorization we found a bug in clang-tidy because `mismatch` got a second range that was too small. It would also give a somewhat better diagnostic message to people misusing algorithms. Maybe something like `__asan_assert_unpoisoned_region` would be best, so people still get the nice ASan output, just with a slightly different failure message.

Reply:
https://github.com/llvm/llvm-project/pull/87665